File: m3-h2/src/test/java/ru/yegorpilipenko/otus/spring/m3h2/controller/BookControllerTest.java
package ru.yegorpilipenko.otus.spring.m3h2.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.val;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import ru.yegorpilipenko.otus.spring.m3h2.dto.ShortBook;
import ru.yegorpilipenko.otus.spring.m3h2.model.Author;
import ru.yegorpilipenko.otus.spring.m3h2.model.Book;
import ru.yegorpilipenko.otus.spring.m3h2.model.Comment;
import ru.yegorpilipenko.otus.spring.m3h2.model.Genre;
import ru.yegorpilipenko.otus.spring.m3h2.service.BookService;

import java.util.List;
import java.util.Optional;

import static org.hamcrest.Matchers.isEmptyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static ru.yegorpilipenko.otus.spring.m3h2.controller.BookController.URL;

@WebMvcTest(BookController.class)
@ExtendWith(SpringExtension.class)
class BookControllerTest {

    @MockBean
    BookService bookService;
    @Autowired
    MockMvc mockMvc;
    ObjectMapper objectMapper = new ObjectMapper();

    @AfterEach
    void clean() {
        // workaround for compatibility @MockBean and @Nested
        reset(bookService);
    }

    @Nested
    @WithMockUser
    class GetAllTest {

        @Test
        void shouldPassForExistsBooks() throws Exception {
            val books = List.of(new ShortBook("1", "n1"), new ShortBook("2", "n2"));
            doReturn(books).when(bookService).getAll();
            mockMvc.perform(get(URL))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                    .andExpect(content().json(objectMapper.writeValueAsString(books)));
            verify(bookService).getAll();
            verifyNoMoreInteractions(bookService);
        }

        @Test
        void shouldPassForNotExistsBooks() throws Exception {
            doReturn(List.of()).when(bookService).getAll();
            mockMvc.perform(get(URL))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                    .andExpect(content().json("[]"));
            verify(bookService).getAll();
            verifyNoMoreInteractions(bookService);
        }

    }

    @Nested
    @WithMockUser
    class GetByIdTest {

        @Test
        void shouldPassForExistsBook() throws Exception {
            val bookId = "1";
            val book = Book.builder()
                    .id(bookId)
                    .name("n")
                    .authors(List.of(new Author("f", "l")))
                    .genres(List.of(new Genre("gn")))
                    .comments(List.of(new Comment("e", "t")))
                    .build();
            doReturn(Optional.of(book)).when(bookService).getById(bookId);
            mockMvc.perform(get("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                    .andExpect(content().json(objectMapper.writeValueAsString(book)));
            verify(bookService).getById(bookId);
            verifyNoMoreInteractions(bookService);
        }

        @Test
        void shouldPassForNotExistsBook() throws Exception {
            val bookId = "1";
            doReturn(Optional.empty()).when(bookService).getById(bookId);
            mockMvc.perform(get("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isNotFound())
                    .andExpect(content().string(isEmptyString()));
            verify(bookService).getById(bookId);
            verifyNoMoreInteractions(bookService);
        }

    }

    @Nested
    @WithMockUser
    class SaveTest {

        @Test
        void shouldPass() throws Exception {
            val book = Book.builder()
                    .id("1")
                    .name("n")
                    .authors(List.of(new Author("f", "l")))
                    .genres(List.of(new Genre("gn")))
                    .comments(List.of(new Comment("e", "t")))
                    .build();
            doReturn(book).when(bookService).save(book);
            mockMvc.perform(
                    post(URL)
                            .contentType(APPLICATION_JSON_UTF8)
                            .content(objectMapper.writeValueAsString(book))
            ).andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                    .andExpect(content().json(objectMapper.writeValueAsString(book)));
        }

    }

    @Nested
    @WithMockUser
    class RemoveByIdTest {

        @Test
        void shouldPassForExistsBook() throws Exception {
            val bookId = "1";
            doReturn(true).when(bookService).removeById(bookId);
            mockMvc.perform(delete("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().string(isEmptyString()));
            verify(bookService).removeById(bookId);
            verifyNoMoreInteractions(bookService);
        }

        @Test
        void shouldPassForNotExistsBook() throws Exception {
            val bookId = "1";
            doReturn(false).when(bookService).removeById(bookId);
            mockMvc.perform(delete("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isNotFound())
                    .andExpect(content().string(isEmptyString()));
            verify(bookService).removeById(bookId);
            verifyNoMoreInteractions(bookService);
        }

    }

}

File: m2-h5/src/test/kotlin/ru/yegorpilipenko/otus/spring/m2h5/controller/BookControllerTest.kt
package ru.yegorpilipenko.otus.spring.m2h5.controller

import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
import com.nhaarman.mockito_kotlin.whenever
import org.assertj.core.api.WithAssertions
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.web.reactive.function.BodyInserters
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ru.yegorpilipenko.otus.spring.m2h5.domain.Book
import ru.yegorpilipenko.otus.spring.m2h5.service.BookService

@WebFluxTest(BookController::class)
@ExtendWith(SpringExtension::class)
class BookControllerTest: WithAssertions {

    @Autowired
    lateinit var webClient: WebTestClient

    @MockBean
    lateinit var bookService: BookService

    @AfterEach
    fun clean() {
        // workaround for nested classes
        Mockito.reset(bookService)
    }

    @Nested
    inner class GetAllTest {

        @Test
        fun shouldPassForExists() {
            val book1 = Book("1", "book1")
            val book2 = Book("2", "book2")
            doReturn(Flux.just(book1, book2)).whenever(bookService).getAll()
            webClient.get().uri("/books").accept(MediaType.APPLICATION_JSON_UTF8)
                    .exchange()
                    .expectStatus().isOk()
                    .expectBodyList(Book::class.java)
                    .hasSize(2)
                    .contains(book1, book2)
            verify(bookService).getAll()
            verifyNoMoreInteractions(bookService)
        }

        @Test
        fun shouldPassForNotExists() {
            doReturn(Flux.empty<Book>()).whenever(bookService).getAll()
            webClient.get().uri("/books")
                    .exchange()
                    .expectStatus().isOk()
            verify(bookService).getAll()
            verifyNoMoreInteractions(bookService)
        }

    }

    @Nested
    inner class GetByIdTest {

        @Test
        fun shouldPassForExists() {
            val id = "1"
            val book = Book(id, "book1")
            doReturn(Mono.just(book)).whenever(bookService).getById(id)
            webClient.get().uri("/books/{id}", book.id).accept(MediaType.APPLICATION_JSON_UTF8)
                    .exchange()
                    .expectStatus().isOk()
                    .expectBody()
                    .jsonPath("$.id").isEqualTo(id)
                    .jsonPath("$.name").isEqualTo(book.name)
            verify(bookService).getById(id)
            verifyNoMoreInteractions(bookService)
        }

        @Test
        fun shouldPassForNotExists() {
            val id = "1"
            doReturn(Mono.empty<Book>()).whenever(bookService).getById(id)
            webClient.get().uri("/books/{id}", id)
                    .exchange()
                    .expectStatus().isOk()
            verify(bookService).getById(id)
            verifyNoMoreInteractions(bookService)
        }

    }

    @Nested
    inner class SaveTest {

        @Test
        fun shouldPass() {
            val id = "1"
            val book = Book(id, "book1")
            doReturn(Mono.just(book)).whenever(bookService).save(book)
            webClient.post().uri("/books").accept(MediaType.APPLICATION_JSON_UTF8).body(BodyInserters.fromObject(book))
                    .exchange()
                    .expectStatus().isOk()
                    .expectBody()
                    .jsonPath("$.id").isEqualTo(id)
                    .jsonPath("$.name").isEqualTo(book.name)
            verify(bookService).save(book)
            verifyNoMoreInteractions(bookService)
        }

    }

    @Nested
    inner class RemoveByIdTest {

        @Test
        fun shouldPass() {
            val id = "1"
            doReturn(Mono.empty<Void>()).whenever(bookService).removeById(id)
            webClient.delete().uri("/books/{id}", id).accept(MediaType.APPLICATION_JSON_UTF8)
                    .exchange()
                    .expectStatus().isOk()
            verify(bookService).removeById(id)
            verifyNoMoreInteractions(bookService)
        }

    }

}
